src/main/java/org/folio/rest/impl/SamlAPI.java

```diff
@@ -185,259 +189,192 @@ public void postSamlCallback(RoutingContext routingContext, Map<String, String>
+        } else {
+          response = PostSamlCallbackResponse.respond500WithTextPlain(cause.getMessage());
         }
+        log.error(cause.getMessage(), cause);
+        asyncResultHandler.handle(Future.succeededFuture(response));
       });
   }
 
-
   @Override
   public void getSamlRegenerate(RoutingContext routingContext, Map<String, String> okapiHeaders,
-                                Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
-
-    regenerateSaml2Config(routingContext)
-      .onComplete(regenerationHandler -> {
-        if (regenerationHandler.failed()) {
-          log.warn("Cannot regenerate SAML2 metadata.", regenerationHandler.cause());
-          String message =
-            "Cannot regenerate SAML2 matadata. Internal error was: " + regenerationHandler.cause().getMessage();
-          asyncResultHandler
-            .handle(Future.succeededFuture(GetSamlRegenerateResponse.respond500WithTextPlain(message)));
-        } else {
-
-          ConfigurationsClient.storeEntry(OkapiHelper.okapiHeaders(okapiHeaders), SamlConfiguration.METADATA_INVALIDATED_CODE, "false")
-            .onComplete(configurationEntryStoredEvent -> {
-
-              if (configurationEntryStoredEvent.failed()) {
-                asyncResultHandler.handle(Future.succeededFuture(GetSamlRegenerateResponse.respond500WithTextPlain("Cannot persist metadata invalidated flag!")));
-              } else {
-                String metadata = regenerationHandler.result();
-
-                Base64Util.encode(vertxContext, metadata)
-                  .onComplete(base64Result -> {
-                    if (base64Result.failed()) {
-                      String message = base64Result.cause() == null ? "" : base64Result.cause().getMessage();
-                      GetSamlRegenerateResponse response = GetSamlRegenerateResponse.respond500WithTextPlain("Cannot encode file content " + message);
-                      asyncResultHandler.handle(Future.succeededFuture(response));
-                    } else {
-                      SamlRegenerateResponse responseEntity = new SamlRegenerateResponse()
-                        .withFileContent(base64Result.result().toString(StandardCharsets.UTF_8));
-                      asyncResultHandler.handle(Future.succeededFuture(GetSamlRegenerateResponse.respond200WithApplicationJson(responseEntity)));
-                    }
-
-                  });
-
-              }
-            });
-        }
+    Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+
+    regenerateSaml2Config(routingContext, vertxContext)
+      .compose(metadata ->
+        ConfigurationsClient.storeEntry(vertxContext.owner(), OkapiHelper.okapiHeaders(okapiHeaders), SamlConfiguration.METADATA_INVALIDATED_CODE, "false")
+          .compose(configurationEntryStoredEvent ->
+            Base64Util.encode(vertxContext, metadata)
```

Comment:
In Base64Util a method `public static String encode(String input)` should be created that avoids the Future, the executeBlocking and the Buffer.

Reply:
sure